src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -9539,16 +9539,28 @@ private bool ScanDesignator()
 
         private bool IsPossibleLambdaExpression(Precedence precedence)
         {
-            if (precedence <= Precedence.Lambda && this.PeekToken(1).Kind == SyntaxKind.EqualsGreaterThanToken)
+            if (precedence <= Precedence.Lambda)
             {
-                return true;
+                SyntaxKind token1 = this.PeekToken(1).Kind;
+                if (token1 == SyntaxKind.EqualsGreaterThanToken)
+                {
+                    return true;
+                }
+                SyntaxKind token2 = this.PeekToken(2).Kind;
+                if (token1 == SyntaxKind.ExclamationToken && token2 == SyntaxKind.EqualsGreaterThanToken)
+                {
+                    return true;
+                }
+                SyntaxKind token3 = this.PeekToken(3).Kind;
```

Comment:
>token3 [](start = 27, length = 6)

Consider inlining this variable since we shouldn't peek ahead unless we need to.